Clear jobs by tag

Zerrossetto opened the pull request:
The intention of this pull is to create an easy way to clear scheduled jobs in a modular environment.
For doing this i added two new methods to `Job` class, `tag()` and `tags()`, that could be used in a way like this one

`schedule.every(10).minutes.do(job).tag('my_tag', 'my_other_tag')`

after doing this it will be possible to clear a subset of jobs by doing

`schedule.clear('my_other_tag')`

always leaving the older way possible by omitting the tag parameter.
Please, tell me if you think it could be useful.

Zerrossetto commented:
Ok, no it's fine. Sorry it was my first experience with CI tools.

laubed commented:
good addition but i think the maintainer is not updating/looking at this repo any more. haven't seen a merge or commit for half a year

Zerrossetto commented:
I'm not an expert but... what is normally done in this cases? The library has a MIT license, so potentially I could go on with my fork.

laubed commented:
Yeah but if everyone thinks that way no progression can be made because 
everyone has it's own repo


Am 24.10.2016 um 09:40 schrieb Duilio Rumbazzi:
>
> I'm not an expert but... what is normally done in this cases? The 
> library has a MIT license, so potentially I could go on with my fork.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub 
> <https://github.com/dbader/schedule/pull/102#issuecomment-255669326>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/ADCLJEgp7cf0wu2cC9WWh9hn84L4-W_Hks5q3GECgaJpZM4KeM7L>.
>

-- 
Daniel Laube | Harzblick 48 | 38122 Braunschweig | +49 176 46669865 | mail@dlaube.de | dlaube.de

Zerrossetto commented:
I agree with you @laubed but currently the author doesn't seem to make the project evolve further. I also tried to contact him by email and waited for 15 days without getting any reply.
Following also your implicit advice I think that I'll get in touch with other project contributors to decide for a single fork to start again implementations on the project.

dbader commented:
Hey thanks for the PR 😄  I can see where this could be useful and I like how it has a relatively small footprint. I'm going to give this a quick code review. I think we should include this in the next release (0.4.0).

In terms of next steps we'd need to:
* resolve code review feedback
* rebase so it builds cleanly against master
* merge

dbader commented on .travis.yml:
```diff
@@ -3,11 +3,13 @@ language: python
-    - pip install coveralls --use-mirrors
+    - pip install pytest-cov
+    - pip install pytest-pep8
+    - pip install coveralls
```
.travis.yml was broken after a pip update on TravisCI. It's fixed in #79, please merge the latest changes from master to resolve.

dbader commented on schedule/__init__.py:
```diff
@@ -75,9 +75,17 @@ def run_all(self, delay_seconds=0):
+            del self.jobs[:]
+        else:
+            old_ref = self.jobs
+            self.jobs = list([job for job in self.jobs if tag not in job.tags])
```
Not sure if I'm missing anything but why call `list()` on the list expression again?

(We might want to turn this into a set however, see below.)

dbader commented on schedule/__init__.py:
```diff
@@ -75,9 +75,17 @@ def run_all(self, delay_seconds=0):
+            self.jobs = list([job for job in self.jobs if tag not in job.tags])
+            global jobs
+            if jobs is old_ref:
+                jobs = self.jobs
```
I'm really starting to hate myself for these module level globals... ;D But I guess for now we'll have to deal with it and patch things up like you're doing here.

dbader commented on schedule/__init__.py:
```diff
@@ -121,6 +129,7 @@ def __init__(self, interval):
         self.next_run = None  # datetime of the next run
         self.period = None  # timedelta between runs, only valid for
         self.start_day = None  # Specific day of the week to start on
+        self.tags = []
```
I think this should be a set so we don't have to deal with duplicate tags.

dbader commented on test_schedule.py:
```diff
@@ -258,3 +258,20 @@ def stop_job():
         schedule.cancel_job(mj)
         assert len(schedule.jobs) == 0
+
+    def test_clear_by_tag(self):
```
Cool, can we add some cases here to:
* ensure invalid tag types (e.g non str/int) raise an exception
* ensure `.tag('tag1', 'tag1', 'tag1')` (reusing the same tag) doesn't cause havoc

Zerrossetto commented on schedule/__init__.py:
```diff
@@ -246,6 +255,10 @@ def sunday(self):
 
+    def tag(self, *tags):
+        self.tags += tags
+        return self
```
I had a thought about this before writing. What I thought was that a tag should actually be anything, also a tuple of tags so I choose to not enforce any kind of tag here.

dbader commented on schedule/__init__.py:
```diff
@@ -75,9 +75,17 @@ def run_all(self, delay_seconds=0):
+            self.jobs = [job for job in self.jobs if tag not in job.tags]
+            global jobs
+            if jobs is old_ref:
+                jobs = self.jobs
```
Here's something to consider: I just realized this changes the current behavior – previously `schedule.jobs` (the module-level global) was only ever modified and not deleted/recreated. So if someone holds a reference to that list and then deletes a tag they're not going to see the updates.

I don't really see how we could modify the jobs list in place in a sensible way though (it's going to be slow). Any thoughts on this?

dbader commented:
The reason why I thought it would make sense to limit allowed types for tags is that one can't just use any type as a job tag – it needs to be comparable and also hashable (because we store them in a set).

The solution I see for this are:
* restrict the types to something sensible that's guaranteed to be comparable and hashable, ie `int` + `str` – downside: it's kind of an arbitrary restriction like you pointed out earlier
* do nothing and add a note to the docstring to mentioning tags need to be comparable and hashable

dbader commented:
Also, could you add an FAQ item that explains the new functionality? We could probably use the PR description for that. Otherwise people won't find this new feature and they'll create more issues :) Thanks!

Zerrossetto commented:
Now, in order, I:

- added Type enforcement for `tag` method that raises a `TypeError` if one of the elements in the arguments list is not hashable
- added a new unittest to test the type enforcement
- added documentation on the FAQs for the method
- corrected the issue over external references over the jobs list
- corrected the FAQs as requested in the roadmap for 0.4.0 #107 

Tell me if I can help in any other way.

dbader commented:
Thanks so much @Zerrossetto, this is great! 😃 Merged.
